idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,371 @@
+        # Fix the estiamted paramerters 
+        for v in theta_names:
+            eval('model.'+var_dic[v]).setlb(theta[v])
+            eval('model.'+var_dic[v]).setub(theta[v])
+    else:
+        for v in theta_names:
+            eval('model.'+v).setlb(theta[v])
+            eval('model.'+v).setub(theta[v])
+    # get gradient of the objective function, constraints, and the column number of each theta
+    gradient_f, gradient_c,line_dic = get_sensitivity(model, theta_names, tee)
+    
+    # calculate error propagation g*cov*g 
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction 
+    if 'gradient_f_theta' in locals():
+        propagation_f['objective'] = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta)))
+   
+
+    # constraints: 
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # save constraints into gradient_cc only if constraints includes theta 
+        # gradient_c[i] is a sparse matrix with column number, row number, and value
+        gradient_cc = []
+        for i in range(gradient_c.shape[0]):
+            if gradient_c[i,[0]] in line_dic.values():
+                gradient_cc.append(gradient_c[i])
+        gradient_cc = np.array(gradient_cc)
+        # save unique constraints numbers that contain theta, index starts from 1. 
+        constriant_number = list(set(gradient_cc[:,[1]].flatten().astype(int)))       
+        # convert sparse matrix to dense matrix with (value, (row number, column number))
+        I,J,V =  gradient_cc[:,[1]].flatten().astype(int)-1, gradient_cc[:,[0]].flatten().astype(int)-1,  gradient_cc[:,[2]].flatten()
+        gradient_cc = sparse.coo_matrix((V,(I,J))).todense()
+        gradient_dic = {}
+        propagation_c = {}
+        
+        # calculate error propagation of constraints
+        for r in constriant_number:
+            gradient_dic['constraints '+str(r)] = gradient_cc[r-1]
+            propagation_c['constraints '+str(r)] = float(np.dot(gradient_cc[r-1],np.dot(cov,np.transpose(gradient_cc[r-1]))))
+    else:
+        propagation_c = {}
+
+    return propagation_f, propagation_c
+
+def get_sensitivity(model, theta_names, tee=False, solver_options=None):
+    """This function calculates gradient vector of the objective function and constraints.
```

Comment:
Mention the gradient vectors are with respect to the variables in `theta_names`.

Reply:
I edited it.